external_services/lti1p3.py

```diff
@@ -0,0 +1,350 @@
+        _add_claim(id_token, "launch_presentation", param)
+
+    def _make_service_claim(self) -> dict:
+        service = {
+            "scope": [
+                    "https://purl.imsglobal.org/spec/lti-ags/scope/lineitem.readonly",
+                    "https://purl.imsglobal.org/spec/lti-ags/scope/score",
+                ],
+            "lineitems": build_aplus_url(api_reverse(
+                    'course-lineitems-list',
+                    kwargs={ 'course_id' : str(self.instance.pk) },
+            )),
+        }
+        if self.exercise:
+            service.update({
+                "lineitem": build_aplus_url(api_reverse(
+                    'course-lineitems-detail',
+                    kwargs={ 'course_id' : str(self.instance.pk), 'exercise_id' : str(self.exercise.pk) },
+                ))
+            })
+        return service
+
+    def finalize_id_token(self, id_token: dict, token: str, token_handler, request) -> str:
+        # Resource link ID should be unique within the platform deployment
+        # Build it based on course instance ID and exercise ID.
+        # If exercise ID is not defined, the link is likely from the course menu.
+        if self.exercise:
+            link_id = str(self.instance.pk) + ':' + str(self.exercise.pk)
+        else:
+            link_id = str(self.instance.pk)
+
+        id_token.update({
+            "iss": build_aplus_url(''),  # SERVICE_BASE_URL or BASE_URL
+            "sub": str(self.user.pk),
+            "exp": id_token["iat"] + 3600,  # one hour
+            "https://purl.imsglobal.org/spec/lti/claim/deployment_id": self.service.deployment_id,
+            "https://purl.imsglobal.org/spec/lti/claim/message_type": "LtiResourceLinkRequest",
+            "https://purl.imsglobal.org/spec/lti/claim/version": "1.3.0",
+            "https://purl.imsglobal.org/spec/lti/claim/resource_link": {
+                "id": link_id,
+            },
+            "https://purl.imsglobal.org/spec/lti/claim/target_link_uri": self.service.url,
+            "https://purl.imsglobal.org/spec/lti/claim/context": {
+                "id": link_id,
+                "title": self.instance.course.name,
+                "type": ["http://purl.imsglobal.org/vocab/lis/v2/course#CourseOffering"]
+            },
+            "https://purl.imsglobal.org/spec/lti-ags/claim/endpoint": self._make_service_claim(),
+        })
+
+        if self.service.share_name:
+            id_token.update({
+                "email": self.user.email,
+                "name": self.user.get_full_name(),
+                "given_name": self.user.first_name,
+                "family_name": self.user.last_name,
+                "https://purl.imsglobal.org/spec/lti/claim/ext": {
+                    "user_username": self.user.username,
+                },
+                "https://purl.imsglobal.org/spec/lti/claim/lis": {
+                    "person_sourcedid": self.user.userprofile.student_id,
```

Comment:
Is this supposed to be unique like `user.id`? `userprofile.student_id` is not unique since different universities may use the same student ids for different people.

Reply:
It is not specifically said how different implementations should use this attribute, and including it is optional, but according to Markus' tool code, MyCourses seems to send student number here. Nevertheless, I now made it "{organization}:{student_id}" for uniqueness. This is now also more similar to the style applied in examples of LTI spec.